### ISSUE TITLE
[BUG] __version__ must be imported first to be used as attr

### ISSUE THREAD

**Issue by nschloe** [pypa/setuptools#3176]
### setuptools version

60.10.0

### Python version

3.9.10

### OS

Ubuntu 22.04

### Description

(The reproduce this bug, [here](https://github.com/nschloe/setuptools-attr-bug) is an MWE.)

In a Python project organized like

```
.
├── pyproject.toml
├── README.md
├── setup.cfg
└── src
    └── foobar
        ├── __about__.py
        ├── __init__.py
        └── _main.py
```
with `__about__.py` reading

```python
__version__ = "1.0.0"
```

I would like to use this `__version__` in `setup.cfg`. I specified

```
version = attr: foobar.__about__.__version__
```

However, this _only_ works if `__init__.py` imports `__about__` before anything else. In the MWE, a third-party library, `rich`, is imported via `main` first, and one gets

```
ModuleNotFoundError: No module named 'rich'
```

when trying to build the project

```
python3 -m build --sdist
```

It's weird that the import order in `__init__.py` should make a difference here.

### Expected behavior

Build correctly independently of the import order.

### How to Reproduce

[See above.](https://github.com/nschloe/setuptools-attr-bug)

### Output

```console
ModuleNotFoundError: No module named 'rich'
```

**Comment by abravalheri** [comment]
Hi @nschloe, thank you very much for reporting this problem.

I have the impression that this issue is solved by the improvements in the `experimental/support-pyproject` branch (planed to be merged soon).

I tested your example locally and it seems to be working for that branch.

If you have time and would like to test your example and original project against this branch, that would be very helpful! You can do that by editing your `pyproject.toml`:

```toml
# replace
requires = ["setuptools>=42", "wheel"]
# with
requires = [
    "setuptools @ git+https://github.com/pypa/setuptools@experimental/support-pyproject",
]
```

(There are more information about other features implemented in this branch in https://discuss.python.org/t/help-testing-experimental-features-in-setuptools/13821)

----
Related question: did this approach used to work in previous versions of `setuptools`?

**Comment by nschloe** [comment]
pyproject files are a different game entirely of course. I'll hold my breath.

**Comment by abravalheri** [comment]
Hi @nschloe you don't have to use pyproject.toml metadata to benefit from the experimental branch 😀

On the process of supporting pyproject.toml metadata, I refactored and improved the code supporting setup.cfg (so the same logic could be shared between both formats).

If you want to test if the changes work for you, you just need to change the `build-system.requires` as I commented above. No need to migrate away from setup.cfg for now.
